### PR TITLE
Solution for the count_numbers_letters challenge

### LINKED ISSUE #35: Counting the number of letters and numbers in an input
## Description

Write a function that counts the number of letters and numbers in a string.

---

## Behavior

The function should:
- Take a string input.
- Returns the count of numbers and letters in the string.
- Give error when the input is not of type string such as none and list.

---

### Example 1

    Input:
    function("hello123")

    Output:
    "Numbers": 3, "Letters": 5, "Others": 0

### Example 2

    Input:
    function("hamid")

    Output:
    "Numbers": 0, "Letters": 5, "Others": 0

### Example 3

    Input:
    function("What?!!!!")

    Output:
    "Numbers": 0, "Letters": 4, "Others": 5

### Example 4

    Input:
    function(["a", 3, "d"])

    Output:
    Error: "Input must not be a list"

### Helpful Links

Here is the [link](https://www.geeksforgeeks.org/python-program-to-calculate-the-number-of-digits-and-letters-in-a-string/) to the challenge on the geeksforgeeks website.

### PR DESCRIPTION
**Pull request by hamid4231**
---
name: solution review 
about: A template PR for code review with a checklist
---

## Behavior

The function should:
- Take a string input.
- Returns the count of numbers and letters in the string.
- Give error when the input is not of type string such as none and list.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [ ] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_count_numbers_letters.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    """
+
+    # Regular Passing Cases
+
+    def test_mixed_characters(self):
+        """
+        Test input containing both numbers and letters.
+        """
+        self.assertEqual(
+            count_numbers_letters("Hello123"), {"Numbers": 3, "Letters": 5, "Others": 0}
```
Comment (ZaidMazen1) [inline solutions/tests/test_count_numbers_letters.py]: I think you should add special characters to your mixed characters test

File: solutions/tests/test_count_numbers_letters.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        )
+
+    def test_only_numbers(self):
+        """
+        Test input containing only numbers.
+        """
+        self.assertEqual(
+            count_numbers_letters("2025"), {"Numbers": 4, "Letters": 0, "Others": 0}
+        )
+
```
Comment (ZaidMazen1) [inline solutions/tests/test_count_numbers_letters.py:58]: Did you forget to add a special characters only test?

File: solutions/tests/test_count_numbers_letters.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    def test_special_characters(self):
+        """
+        Test input containing special characters.
+        """
+        self.assertEqual(
+            count_numbers_letters("@123!aB"), {"Numbers": 3, "Letters": 2, "Others": 2}
+        )
+
```
Comment (ZaidMazen1) [inline solutions/tests/test_count_numbers_letters.py:76]: Important edge case could be an input with exactly 100 characters

File: solutions/count_numbers_letters.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    """
+
+    # Define a maximum allowed length for the input
+    MAX_LENGTH = 100
+
+    # Validating User Inputs
+    if user_input is None:
+        raise ValueError("Input must not be None.")
+
+    if isinstance(user_input, list):
```
Comment (ZaidMazen1) [inline solutions/count_numbers_letters.py:45]: Maybe I don't understand code properly but you have a test to make sure it is not a list and another test to make sure it is a string. Could line 44 and 45 be deleted?

File: solutions/count_numbers_letters.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+"""
+This module provides a function to find the number of
+letters, numbers and other characters in a given string.
+Created on: 2025/1/6
+@author: Hamidullah Rajabi
+"""
+
+
+def count_numbers_letters(user_input):
+    """
```
Comment (ZaidMazen1) [inline solutions/count_numbers_letters.py:13]: def count_numbers_letters(user_input: str):

would be a bit more clear to indicate that it must be a string

**Comment by ZaidMazen1** [comment]
Hey Hamid. Great work. I've just made a couple comments that should be quick and easy to resolve. One more thing is that you've made a function that counts letters, numbers, and special characters. But from the unittest and function name, it seems to me that you treat special characters as edge cases. Perhaps either clarify this in the docstring(s) or rename the function to reflect your function more appropriately.

e.g count_char_type

File: solutions/count_numbers_letters.py (hunk 5 above)
Comment (hamid4231) [reply solutions/count_numbers_letters.py:13]: noted, will specify the input type.

File: solutions/count_numbers_letters.py (hunk 4 above)
Comment (hamid4231) [reply solutions/count_numbers_letters.py:45]: yes actually, one of the defensive test cases is a list whether it is list of integers or strings. and the other defensive test case is non string which can be type of nan, none, integers, or floats.

File: solutions/tests/test_count_numbers_letters.py (hunk 1 above)
Comment (hamid4231) [reply solutions/tests/test_count_numbers_letters.py]: sure, will make one for special characters.

File: solutions/tests/test_count_numbers_letters.py (hunk 3 above)
Comment (hamid4231) [reply solutions/tests/test_count_numbers_letters.py:76]: since I have a couple of edge and defensive cases, I thought it would be enough and didn't think of more.

**Review by ZaidMazen1 — APPROVED** [review]
Nice changes, good work Hamid. Approved the solution now.
